File: app/src/main/java/com/doubean/ford/util/LiveDataUtil.java
package com.doubean.ford.util;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class LiveDataUtil {

    public static <T> LiveData<List<T>> zip(LiveData<T>... liveItems) {
        final List<T> zipped = new ArrayList<>(Arrays.asList((T[]) new Object[liveItems.length]));
        final MediatorLiveData<List<T>> mediator = new MediatorLiveData<>();
        for (int i = 0; i < liveItems.length; i++) {
            int finalI = i;
            mediator.addSource(liveItems[i], o -> {
                zipped.set(finalI, o);
                mediator.setValue(zipped);
            });
        }
        return mediator;
    }
}

File: app/src/main/java/com/doubean/ford/util/LiveDataUtil.kt
package com.doubean.ford.util

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData

//TODO: use Flow instead of LiveData to process complex data logic
object LiveDataUtil {
    inline fun <reified T> zip(vararg liveItems: LiveData<T>): LiveData<List<T>> {
        val zipped = arrayOfNulls<T>(liveItems.size)
        val mediator = MediatorLiveData<List<T>>()
        for (i in liveItems.indices) {
            mediator.addSource(liveItems[i]) { o: T ->
                zipped[i] = o
                if (!zipped.contains(null)) {
                    val value = List<T>(liveItems.size) { zipped[i]!! }
                    mediator.value = value
                }
            }
        }
        return mediator
    }
}